File: JavaScript/7js 1/arrayLeaders.js
/*
|--------------------------------------------------------------------------
| Array Leaders (Array Series #3)
|--------------------------------------------------------------------------
|
*/

function arrayLeaders(numbers) {
   return numbers.filter((v, i) => v > numbers.slice(i + 1).reduce((a, b) => a + b, 0));
}


const arrayLeaders = (n, s = eval(n.join`+`)) => n.filter((v) => v > (s -= v));


function arrayLeaders(numbers) {
   let arr = [];
   for (i = 0; i < numbers.length; i++) {
      if (numbers[i] > numbers.slice(i + 1).reduce((a, b) => a + b, 0)) arr.push(numbers[i]);
   }
   return arr;
}


function arrayLeaders(numbers) {
   return numbers.reduceRight(
      ([acc, sum], x) => (x > sum ? [[x, ...acc], sum + x] : [acc, sum + x]),
      [[], 0]
   )[0];
}


console.log(arrayLeaders([1, 2, 3, 4, 0]), [4]);
console.log(arrayLeaders([16, 17, 4, 3, 5, 2]), [17, 5, 2]);

File: Python/7py 1/array_leaders.py
"""
|--------------------------------------------------------------------------
| Array Leaders (Array Series #3)
|--------------------------------------------------------------------------
|
"""

def array_leaders(numbers):
   return [v for i, v in enumerate(numbers) if v > sum(numbers[(i+1):])]


def array_leaders(n):
   res = []
   for i in range(len(n)):
      if n[i] > sum(n[i+1:]): res.append(n[i])
   return res


def array_leaders(arr):
   res = []
   while arr:
      temp = arr.pop(0)
      if temp > sum(arr): res.append(temp)
   return res


print(array_leaders([1,2,3,4,0]), [4])
print(array_leaders([16,17,4,3,5,2]), [17,5,2])